clients/src/main/java/org/apache/kafka/clients/producer/internals/StickyPartitionCache.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.PartitionInfo;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.kafka.common.utils.Utils;
+
+/**
+ * An internal class that implements a cache used for sticky partitioning behavior. The cache tracks the current sticky
+ * partition for any given topic. This class should not be used externally. 
+ */
+public class StickyPartitionCache {
+    private final ConcurrentMap<String, Integer> indexCache;
+    public StickyPartitionCache() {
+        this.indexCache = new ConcurrentHashMap<>();
+    }
+
+    public int partition(String topic, Cluster cluster) {
+        Integer part = indexCache.get(topic);
+        if (part == null) {
+            return nextPartition(topic, cluster, -1);
+        }
+        return part;
+    }
+
+    public int nextPartition(String topic, Cluster cluster, int prevPartition) {
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        int numPartitions = partitions.size();
+        Integer oldPart = indexCache.get(topic);
+        Integer newPart = oldPart;
+        if (oldPart == null || oldPart == prevPartition) {
```

Comment:
Would it be worth commenting on the reason for the `oldPart == prevPartition` check? I'm guessing this is due to another send switching the partition out from underneath it, meaning that we should simply get the partition that exists in the cache?

Reply:
It is that, or potentially we could have a keyed value triggering a new batch when we do not need a new sticky batch. Checking the prevPartition would help in this case too.

Reply:
_// Check that the current sticky partition for the topic is either not set or that the partition that just filled up matches the sticky partition that needs to be changed._

Technically this comment does not cover one case but I don't know if it's worth being overly pedantic in the comment to cover all cases.

Reply:
Actually this might be better:
_// Check that the current sticky partition for the topic is either not set or that the partition that triggered the new batch matches the sticky partition that needs to be changed._